apps/prairielearn/src/pages/elementFiles/elementFiles.js

```diff
@@ -18,50 +22,74 @@ const router = Router({ mergeParams: true });
-    // for backwards compatibility. See `server.js` for more details.
-    //
-    // As with `/assets/`, we assume that element files are likely to change
-    // when running in dev mode, so we skip caching entirely in that case.
-    const isCached = !!req.params.cachebuster && !config.devMode;
-    const sendFileOptions = {
-      immutable: isCached,
-      maxAge: isCached ? '31536000s' : 0,
-    };
+      // If the route includes a `cachebuster` param, we'll set the `immutable`
+      // and `maxAge` options on the `Cache-Control` header. This router is
+      // mounted twice - one with the cachebuster in the URL, and once without it
+      // for backwards compatibility. See `server.js` for more details.
+      //
+      // As with `/assets/`, we assume that element files are likely to change
+      // when running in dev mode, so we skip caching entirely in that case.
+      const isCached = !!req.params.cachebuster && !config.devMode;
+      const sendFileOptions = {
+        immutable: isCached,
+        maxAge: isCached ? '31536000s' : 0,
+      };
 
-    if (isCached) {
-      // `middlewares/cors.js` disables caching for all routes by default.
-      // We need to remove this header so that `res.sendFile` can set it
-      // correctly.
-      res.removeHeader('Cache-Control');
-    }
+      if (isCached) {
+        // `middlewares/cors.js` disables caching for all routes by default.
+        // We need to remove this header so that `res.sendFile` can set it
+        // correctly.
+        res.removeHeader('Cache-Control');
+      }
 
-    let elementFilesDir;
-    if (res.locals.course) {
-      // Files should be served from the course directory
-      const coursePath = chunks.getRuntimeDirectoryForCourse(res.locals.course);
-      await chunks.ensureChunksForCourseAsync(res.locals.course.id, { type: 'elements' });
-      elementFilesDir = path.join(coursePath, 'elements');
-      res.sendFile(filename, { root: elementFilesDir, ...sendFileOptions });
-    } else {
-      elementFilesDir = path.join(APP_ROOT_PATH, 'elements');
-      res.sendFile(filename, { root: elementFilesDir, ...sendFileOptions });
-    }
-  }),
-);
+      let elementFilesDir;
+      if (options.publicEndpoint && !options.static) {
+        const variant = await selectVariantById(z.string().parse(req.query.variant_id));
+        if (!variant) throw new HttpStatusError(404, 'Not Found');
+        const question = await selectQuestionById(variant.question_id);
+        if (!question.shared_publicly) throw new HttpStatusError(404, 'Not Found');
+        const course = await selectCourseById(req.params.course_id);
+        const coursePath = chunks.getRuntimeDirectoryForCourse(course);
+        await chunks.ensureChunksForCourseAsync(course.id, { type: 'elements' });
+        elementFilesDir = path.join(coursePath, 'elements');
+      } else if (!options.publicEndpoint && !options.static) {
+        // Files should be served from the course directory
```

Comment:
This is the actual new code in this PR, the rest is a result of indentation changes.